subsys/bluetooth/mesh/net.c

```diff
@@ -851,6 +852,10 @@ void bt_mesh_net_recv(struct net_buf_simple *data, int8_t rssi,
 		return;
 	}
 
+	if (IS_ENABLED(CONFIG_BT_MESH_STATISTIC)) {
+		bt_mesh_stat_rx(net_if);
+	}
```

Comment:
```suggestion
#ifdef CONFIG_BT_MESH_STATISTIC
	bt_mesh_stat_rx(net_if);
#endif
```

Reply:
Both variants are the same. However, usage of `IS_ENABLED` is preferred. The preprocessor just cuts out the source code and if there is syntax (or other) issues they are hidden from the compiler. Usage of `IS_ENABLED` macro allows to catch them and keep source code in the better shape.
It is well described here: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/util_macro.h#L96-L124

Reply:
Are they the same? Isn't this an issue for the linker as `bt_mesh_stat_rx` will not be compiled in the sources?

Reply:
Hm... which issue? All twister steps passed.

Reply:
You're relying on [Dead code elimination](https://en.wikipedia.org/wiki/Dead-code_elimination) done by the compiler. I'm not sure you can always do this, as it depends on which compiler and moreover which optimization you are using.

The `bt_mesh_stat_rx` function won't be found by the linker because `statistic.c` is only compiled if `CONFIG_BT_MESH_STATISTIC` is defined.

Try to compile with `CONFIG_BT_MESH_STATISTIC` disabled and no compiler optimization, does this still build?

Reply:
I'm compiling the code with `-O0` and don't get warning.

In any case, I think we gain more from using `IS_ENABLED` for cases like this one than with `#if-#endif`. We don't have full coverage of all features. This allows us to catch issues where we forgot to updated code. And I think this pattern is used in many places in Zephyr.

Reply:
Or, as mentioned [here](https://github.com/zephyrproject-rtos/zephyr/pull/59962#pullrequestreview-1512475731), we can remove the `IS_ENABLED/#if-#endif` all together.

Reply:
This then should be used everywhere, but I'm not sure we need to change this pattern in this particular PR.

Reply:
Tried all available in Zephyr optimization options: https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/compiler/gcc/compiler_flags.cmake#L15-L22  for arm and host gcc.
Dead code is always eliminated. 
I was able to catch the issue only when applied specific optimization to exact function using attributes. I guess, a customer should understand consequences when applies manual region code optimization.

Seems this is the topic about the problem that doesn't exist.

Reply:
Just for the note. Optimization options can set through menuconfig in `Build and Link Features` -> `Compiler Options` -> `Optimization level`. Additional compiler options can be manually added in `Custom compiler options` option in the same menu.